chia/full_node/coin_store.py

```diff
@@ -411,6 +411,83 @@ async def get_coin_states_by_ids(
+
+        coin_states: List[CoinState] = []
+
+        async with self.db_wrapper.reader_no_transaction() as conn:
+            puzzle_hashes_db = tuple(puzzle_hashes)
+            puzzle_hash_count = len(puzzle_hashes_db)
+
+            if include_hinted:
+                cursor = await conn.execute(
+                    f"SELECT cr.confirmed_index, cr.spent_index, cr.coinbase, cr.puzzle_hash, "
+                    f"cr.coin_parent, cr.amount, cr.timestamp FROM coin_record cr "
+                    f"LEFT JOIN hints h ON cr.coin_name = h.coin_id "
+                    f'WHERE (cr.puzzle_hash in ({"?," * (puzzle_hash_count - 1)}?) '
+                    f'OR h.hint in ({"?," * (puzzle_hash_count - 1)}?)) '
+                    f"AND (cr.confirmed_index>=? OR cr.spent_index>=?) "
+                    f"{'' if include_spent else 'AND cr.spent_index=0'} "
+                    f"{'' if include_unspent else 'AND cr.spent_index>0'} "
+                    f"ORDER BY MAX(cr.confirmed_index, cr.spent_index) ASC "
+                    f"LIMIT ?",
+                    puzzle_hashes_db + puzzle_hashes_db + (min_height, min_height, max_items + 1),
+                )
+            else:
+                cursor = await conn.execute(
+                    f"SELECT confirmed_index, spent_index, coinbase, puzzle_hash, "
+                    f"coin_parent, amount, timestamp FROM coin_record INDEXED BY coin_puzzle_hash "
+                    f'WHERE puzzle_hash in ({"?," * (puzzle_hash_count - 1)}?) '
+                    f"AND (confirmed_index>=? OR spent_index>=?) "
+                    f"{'' if include_spent else 'AND spent_index=0'} "
+                    f"{'' if include_unspent else 'AND spent_index>0'} "
+                    f"ORDER BY MAX(confirmed_index, spent_index) ASC "
+                    f"LIMIT ?",
+                    puzzle_hashes_db + (min_height, min_height, max_items + 1),
+                )
+
+            for row in await cursor.fetchall():
+                coin_states.append(self.row_to_coin_state(row))
+
+            is_finished: bool
+
+            if len(coin_states) == max_items:
+                next_height = min_height
+                is_finished = True
+            elif len(coin_states) > max_items:
+                next_coin_state = coin_states.pop()
+                next_height = uint32(max(next_coin_state.created_height or 0, next_coin_state.spent_height or 0))
+
+                while len(coin_states) > 0:
+                    last_coin_state = coin_states[-1]
+                    height = uint32(max(last_coin_state.created_height or 0, last_coin_state.spent_height or 0))
+                    if height == min_height:
+                        coin_states.pop()
+                    else:
+                        break
+
+                is_finished = False
+            else:
+                next_height = min_height
+                is_finished = True
+
+        return (coin_states, next_height, is_finished)
```

Comment:
you don't technically need the parentheses in the `return` statement

Reply:
The more you know :P

Reply:
The comma makes the tuple...  except in the case of an empty tuple where the parentheses do.  Or as a function parameter where the parentheses clarify the commas are not delimiting function arguments.  See, totally simple...  `:]`